### PR TITLE
[PATCH v6] Checksums generation support

### PR DESCRIPTION
**Pull request by lumag**
Smaller version of #389 . Removed software implementation of checksum validation support.

### CODE REVIEW

File: platform/linux-generic/pktio/loop.c [hunk 1]
```diff
@@ -172,6 +175,41 @@ static int loopback_recv(pktio_entry_t *pktio_entry, int index ODP_UNUSED,
...
+			      uint8_t *l4_proto)
+{
+	uint8_t l3_proto_ver = _ODP_IPV4HDR_VER(*(uint8_t *)l3_hdr);
+
+	if (l3_proto_ver == _ODP_IPV4 && l3_len >= _ODP_IPV4HDR_LEN) {
+		_odp_ipv4hdr_t *ip = l3_hdr;
+
+		*l3_proto_v4 = 1;
+		if (!_ODP_IPV4HDR_IS_FRAGMENT(_odp_be_to_cpu_16(ip->frag_offset)))
+			*l4_proto = ip->proto;
```
Comment (Bill-Fischofer-Linaro) [inline platform/linux-generic/pktio/loop.c]: Checkpatch flags several line-too-long errors for this patch:
```
WARNING: line over 80 characters
#58: FILE: platform/linux-generic/pktio/loop.c:192:
+		if (!_ODP_IPV4HDR_IS_FRAGMENT(_odp_be_to_cpu_16(ip->frag_offset)))
WARNING: line over 80 characters
#113: FILE: platform/linux-generic/pktio/loop.c:279:
+						   pkt_hdr->p.flags.l3_chksum_set,
WARNING: line over 80 characters
#118: FILE: platform/linux-generic/pktio/loop.c:284:
+						   pkt_hdr->p.flags.l4_chksum_set,
WARNING: line over 80 characters
#123: FILE: platform/linux-generic/pktio/loop.c:289:
+						   pkt_hdr->p.flags.l4_chksum_set,
total: 0 errors, 4 warnings, 0 checks, 155 lines checked
NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.
0005-linux-gen-pktio-loop-support-IPv4-TCP-UDP-checksum-g.patch has style problems, please review.
```

File: platform/linux-generic/odp_ipsec.c [hunk 2]
```diff
@@ -1295,6 +1296,40 @@ static void ipsec_out_ah_post(ipsec_state_t *state, odp_packet_t pkt)
...
+static void ipsec_out_checksums(odp_packet_t pkt,
+				ipsec_state_t *state)
+{
+	odp_bool_t ipv4_chksum_pkt, udp_chksum_pkt, tcp_chksum_pkt;
+	odp_packet_hdr_t *pkt_hdr = packet_hdr(pkt);
+
+	ipv4_chksum_pkt = OL_TX_CHKSUM_PKT(ipsec_config.outbound.chksum.inner_ipv4,
+					   state->is_ipv4,
+					   pkt_hdr->p.flags.l3_chksum_set,
+					   pkt_hdr->p.flags.l3_chksum);
```
Comment (Bill-Fischofer-Linaro) [inline platform/linux-generic/odp_ipsec.c]: Checkpatch flags several line-too-long errors for this patch:
```
WARNING: line over 80 characters
#36: FILE: platform/linux-generic/odp_ipsec.c:1308:
+	ipv4_chksum_pkt = OL_TX_CHKSUM_PKT(ipsec_config.outbound.chksum.inner_ipv4,
WARNING: line over 80 characters
#40: FILE: platform/linux-generic/odp_ipsec.c:1312:
+	udp_chksum_pkt =  OL_TX_CHKSUM_PKT(ipsec_config.outbound.chksum.inner_udp,
WARNING: line over 80 characters
#45: FILE: platform/linux-generic/odp_ipsec.c:1317:
+	tcp_chksum_pkt =  OL_TX_CHKSUM_PKT(ipsec_config.outbound.chksum.inner_tcp,
total: 0 errors, 3 warnings, 0 checks, 64 lines checked
NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.
0007-linux-gen-ipsec-implement-outbound-checksumming-supp.patch has style problems, please review.
```

**Comment by bogdanPricope** [comment]
Looks reasonable.

File: platform/linux-generic/include/odp_chksum_internal.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+/**
+ * @file
+ *
+ * ODP checksum - implementation internal
+ */
```
Comment (psavol) [inline platform/linux-generic/include/odp_chksum_internal.h]: Internal files are not included into doxygen documentation. So, these 5 lines of doxygen documentation are unnecessary and can be removed.

Comment (lumag) [reply platform/linux-generic/include/odp_chksum_internal.h]: ack

File: platform/linux-generic/pktio/loop.c (hunk 1 above)
Comment (lumag) [reply platform/linux-generic/pktio/loop.c]: Fixed both files.

File: platform/linux-generic/odp_chksum.c [hunk 4]
```diff
@@ -7,26 +7,14 @@
...
-	}
-
-	/* Add left-over byte, if any */
-	if (len > 0) {
-		uint16_t left_over = 0;
-
-		*(uint8_t *)&left_over = *(const uint8_t *)data;
-		sum += left_over;
-	}
+	uint32_t sum = _odp_chksum_ones_comp16_32(p, len, false);
```
[the remark this answers is not in the document]
Comment (lumag) [reply platform/linux-generic/odp_chksum.c]: I'll have to think about it.

File: platform/linux-generic/include/odp_chksum_internal.h [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ * Based on RFC1071 and its errata.
+ */
+typedef union {
+	uint16_t w;
+	uint8_t b[2];
+} swap_buf_t;
+
+static uint32_t _odp_chksum_ones_comp16_32(const uint8_t *p, uint32_t len,
+					   odp_bool_t odd_offset)
+
```
Comment (psavol) [inline platform/linux-generic/include/odp_chksum_internal.h]: odd_offset and p%2 seem overlapping checks. Couldn't this function always check and move to an even address and then use the API function for the rest of the (now even address) calculation, and combine results in the end.

Pseudo code:

```
uint32_t odd_checksum(uint8_t *p, uint32_t len)
{
    if (p%2) {
        sw.b[0] = 0;
        sw.b[1] = *(p++);
        odd_sum = sw.w;
        len--;
    }

    even_sum = odp_chksum_ones_comp16(p, len);

    return odd_sum + even_sum;
}
```

Comment (lumag) [reply platform/linux-generic/include/odp_chksum_internal.h]: @psavol Maybe I should add this as an in-source comment. There is difference between in-packet offset oddity and data address offset oddity. As I said, I'll think about improving this low-level function. Compare it with helper function, which also has this odd/even in-packet support, but always uses `uint16_t` to further access data.

**Comment by lumag** [comment]
@psavol I have updated PR to move the internal worker function into `odp_packet.c` and expanded comments/parameters there, so that there is no more `odd_offset`.

File: platform/linux-generic/odp_packet.c [hunk 6]
```diff
@@ -1910,6 +1910,92 @@ int _odp_packet_copy_md_to_packet(odp_packet_t srcpkt, odp_packet_t dstpkt)
...
+
+{
+	uint32_t sum = 0;
+
+	/* Include second part of 16-bit short word split between segments */
+	if (len > 0 && (offset % 2)) {
+		swap_buf_t sw;
+
+		sw.b[0] = 0;
+		sw.b[1] = *(p++);
```
Comment (Bill-Fischofer-Linaro) [inline platform/linux-generic/odp_packet.c]: Not wrong, but unnecessary as `*p++` is equivalent. See [here](https://www.geeksforgeeks.org/difference-between-p-p-and-p/) for a discussion. Same comment for other instances of this idiom.

Comment (lumag) [reply platform/linux-generic/odp_packet.c]: Fixed now

**Comment by muvarov** [comment]
Merged.

**Pull request closed without merging**